src/routes_api.py

```diff
@@ -51,4 +57,204 @@ def trigger_scheduled_run(cid, aid, scheduled_run_id):
                 db.update_scheduled_run_status(sched_run["_id"], ScheduledRunStatus.RAN)
                 db.update_scheduled_run_bw_run_id(sched_run["_id"], bw_run_id)
             return util.success("")
+        
+        # Want to avoid stuff like this, with overlaps in function definitions
+        # Best way is to consider an AdminOperations class and have AdminRoutes and APIRoutes
+        # use the functionality defined in there, instead of whatever I did with AdminRoutes currently
+        @blueprint.route("/api/<cid>/<aid>/add_extension", methods=["POST"])
+        @auth.require_course_auth
+        @auth.require_admin_status
+        def add_extension(cid, aid):
+            form = request.json
+            
+            assignment = db.get_assignment(cid, aid)
+            if not assignment:
+                return util.error("Invalid course or assignment.\nPlease try again.")
+            
+        
+            missing = util.check_missing_fields(form, "netids", "max_runs", "start", "end")
+            if missing:
+                return util.error(f"Extension missing fields {', '.join(missing)}. Please try again.")
+
+            student_netids = form["netids"].replace(" ", "").lower().split(",")
+            for student_netid in student_netids:
+                if not util.valid_id(student_netid) or not verify_student(student_netid, cid):
+                    return util.error(f"Invalid or non-existent student NetID: {student_netid}")
+
+            try:
+                max_runs = int(form["max_runs"])
+                if max_runs < 1:
+                    return util.error("Max Runs must be a positive integer.")
+            except ValueError:
+                return util.error("Max Runs must be a positive integer.")
+
+            # print(form["start"], form["end"])
+
+            start = util.parse_form_datetime(form["start"])
+            if not start:
+                return util.error("Failed to parse timestamp")
+            start = start.timestamp()
+            end = util.parse_form_datetime(form["end"])
+            if not end:
+                return util.error("Failed to parse timestamp")
+            end = end.timestamp()
+            if start >= end:
+                return util.error("Start must be before End")
+
+            ext_res = db.add_extension(cid, aid, ','.join(student_netids), max_runs, start, end)
+            if not ext_res.acknowledged:
+                return util.error("Failed to add extension to db")
+            
+            form = request.json
```

Comment:
This looks like a duplicate of line 68 to me unless I'm missing something.

Reply:
I assume you mean 108? Fixed.